extensions/scheduler/common/src/main/java/io/quarkus/scheduler/common/runtime/DefaultInvoker.java

```diff
@@ -14,11 +15,18 @@ public CompletionStage<Void> invoke(ScheduledExecution execution) throws Excepti
         if (requestContext.isActive()) {
             return invokeBean(execution);
         } else {
+            // 1. Activate the context
+            // 2. Store the state (which is basically a shared Map instance)
+            // 3. Destroy the context correctly when the returned stage completes
+            requestContext.activate();
+            final ContextState state = requestContext.getState();
             try {
-                requestContext.activate();
-                return invokeBean(execution);
+                return invokeBean(execution).whenComplete((v, t) -> {
+                    requestContext.destroy(state);
+                });
             } finally {
-                requestContext.terminate();
+                // Always deactivate the context
+                requestContext.deactivate();
```

Comment:
Are you sure this is right?

The context could be deactivated before being used by `invokeBean`.

Reply:
Yes, _deactivation_ means that we remove a threadlocal but do not destroy the context/beans, i.e. the thread is "clear". If Mutiny is used inside the `invokeBean()` then we rely on context propagation to make a snapshot of the current context and propagate it accordingly.

It wouldn't work for `CompletionStage` with a custom executor though. In this case, a user needs to handle the context propagation manually (as it was before).

Reply:
Looks correct, CP takes snapshot based on the `ContextState` of currently active context when it is captured. If I get this scenario right, then this will happen when you perform `invokeBean()`.